ndnsd/discovery/service-discovery.cpp

```diff
@@ -0,0 +1,374 @@
+    dataContent += "|";
+    dataContent += item.first;
+    dataContent += "|";
+    dataContent += item.second;
+  }
+  return dataContent;
+}
+
+void
+ServiceDiscovery::sendData(const ndn::Name& name)
+{
+  NDNSD_LOG_INFO("Sending data for: " << name);
+  auto timeDiff = ndn::time::system_clock::now() - m_producerState.publishTimestamp;
+  auto status = (timeDiff >= m_producerState.serviceLifetime*1000)
+                          ? EXPIRED : ACTIVE;
+
+  std::shared_ptr<ndn::Data> replyData = std::make_shared<ndn::Data>(name);
+  replyData->setFreshnessPeriod(1_ms);
+
+  auto dataContent = makeDataContent();
+  auto& data = wireEncode(dataContent, status);
+  replyData->setContent(data);
+  m_keyChain.sign(*replyData);
+
+  try
+  {
+    m_face.put(*replyData);
+  }
+  catch (const std::exception& ex) {
+    std::cerr << ex.what() << std::endl;
+  }
+}
+
+void
+ServiceDiscovery::expressInterest(const ndn::Name& name)
+{
+  ndn::Interest interest(name);
+  interest.setCanBePrefix(false);
+  interest.setMustBeFresh(true);
+  interest.setInterestLifetime(160_ms);
+
+  m_face.expressInterest(interest,
+                         bind(&ServiceDiscovery::onData, this, _1, _2),
+                         bind(&ServiceDiscovery::onTimeout, this, _1),
+                         bind(&ServiceDiscovery::onTimeout, this, _1));
+
+}
+
+void
+ServiceDiscovery::onData(const ndn::Interest& interest, const ndn::Data& data)
+{
+  data.getContent().parse();
+  auto consumerReply = wireDecode(data.getContent().get(tlv::DiscoveryData));
+  m_discoveryCallback(consumerReply);
+  m_counter--;
+
+  if (m_counter <= 0)
+    stop();
+}
+
+void
+ServiceDiscovery::onTimeout(const ndn::Interest& interest)
+{
+  if (m_appType == CONSUMER)
+  {
+    m_counter--;
+  }
+  NDNSD_LOG_INFO("Interest: " << interest.getName() << "timeout");
+}
+
+void
+ServiceDiscovery::registrationFailed(const ndn::Name& name)
+{
+  NDNSD_LOG_ERROR("ERROR: Failed to register prefix " << name << " in local hub's daemon");
+}
+
+void
+ServiceDiscovery::onRegistrationSuccess(const ndn::Name& name)
+{
+  NDNSD_LOG_DEBUG("Successfully registered prefix: " << name);
+}
+
+void
+ServiceDiscovery::doUpdate(const ndn::Name& prefix)
+{
+  m_syncAdapter.publishUpdate(prefix);
+  NDNSD_LOG_INFO("Publish: " << prefix);
+}
+
+void
+ServiceDiscovery::processSyncUpdate(const std::vector<ndnsd::SyncDataInfo>& updates)
+{
+  m_counter = updates.size();
+  if (m_appType == CONSUMER)
+  {
+    for (auto item: updates)
+    {
+      NDNSD_LOG_INFO("Fetching data for prefix:" << item.prefix);
+      expressInterest(item.prefix);
+    }
+  }
+  else
+  {
+    Reply consumerReply;
+    consumerReply.serviceDetails.insert(std::pair<std::string, std::string>("prefix", m_producerState.applicationPrefix.toUri()));
+    consumerReply.status = ACTIVE;
+    m_discoveryCallback(consumerReply);
+  }
+}
+
+template<ndn::encoding::Tag TAG>
+size_t
+ServiceDiscovery::wireEncode(ndn::EncodingImpl<TAG>& encoder,
```

Comment:
Never seen a wireEncode like this. Usually info and status comes from a class member.
If this is the case then you are right, in this program having m_wire as a class member does not make sense either.

Reply:
Yes, this wireEncode is a bit weird. I kept m_wire member variable because I need to return wire from a wireEncode which will throw "a reference to stack memory associated with local variable 'wire' returned". Either I have to dynamically allocate memory for "wire" or pass wire to wireEncode from callee. 

Any suggestions?